elastic/beats
# Restart Beat when certificates or key files are modified

## What does this PR do?

This PR enables any Beat to reexec when a output CA certificate file is modified. The main use case we're covering here is enabling "CA certificates reload" when running Metricbeat on ECK to monitor Elasticsearch. The "Related issues" section contains links with more details.

`libbeat/management/management.go` is also refactored and some unused bits of code are removed.

## Why is it important?

It enables Beats to re-start whenever the CA certificate for an output changes. This enables the Metricbeat side car used by ECK to monitor Elasticsearch to "reload" the certificates without having to restart Elasticsearch container.

## Checklist


## Author's Checklist

- [x] Proper names for the loggers
- [x] Should we rename the `nilManager`? It has never been `nil`, and now it has got some function.
- [x] Fix config keys and docs
- [x] Finish Windows implementation
- [x] Add documentation and logs about the supported OSes
- [x] Document about the needed seccomp rule to allow the reexec syscall
- [x] Test on MacOS
- [x] Update PR description with restart instead of exit, etc.

## How to test this PR locally

1. Run any Beat with an elasticsearch output `ssl` section like this:
    ```yaml
      ssl:
        certificate_authorities: ["/path/to/your/ca/cert.pem"]
        verification_mode: full
        restart_on_cert_change:
          enabled: true
          period: 1m
    ```
3. Wait for the Beat to start
4. Touch one of the files defined on `output.elasticsearch.ssl.certificate_authorities`
5. The Beat restart. You can see the shutdown/start logs

You can also run the test defined on my branch: https://github.com/belimawr/beats/tree/reload-certificates-key-files-tests-2. Because of some global state, only one test from `beat_integration_test.go` can be run, hence it is not part of this PR.

1. Go to `beats/libbeat/cmd/instance`
2. Run the test with `go test -v -run="TestExit" -timeout=40s`
3. The test should pass.


## Related issues
- Relates #15956
- Closes #34408

## Use cases

- Reloading certificates on ECK

~~## Screenshots~~
~~## Logs~~
